src/Dropdown/Dropdown.js

```diff
@@ -147,67 +159,206 @@ export default class Dropdown extends Component<Props, State> {
-      'aria-activedescendant': isOpen
-        ? this.selectedItemId || `${contentId}-menu`
-        : undefined,
-      'aria-haspopup': true,
-      contentId,
-      onKeyDown: this.onTriggerKeyDown
-    };
+  getContentId = () => {
+    return `${this.id}-content`;
+  };
+
+  getMenuId = () => {
+    return `${this.id}-menu`;
+  };
+
+  getMenuItemId = (index: string) => {
+    return `${this.id}-item-${index}`;
+  };
+
+  getTriggerProps = (props: Object = {}) => {
+    const contentId = this.getContentId();
+    const isOpen = this.getControllableValue('isOpen');
+
+    return composePropsWithGetter(
+      {
+        // Props set by caller, e.g. Popover
+        ...props,
+
+        // Props set by this component
+        'aria-activedescendant': isOpen
+          ? this.getHighlightedItemId() || this.getMenuId()
+          : undefined,
+        'aria-describedby': contentId,
+        'aria-haspopup': true,
+        'aria-owns': contentId,
+        onKeyDown: this.onTriggerKeyDown,
+        onKeyUp: (event: SyntheticKeyboardEvent<>) => {
+          // Prevent Firefox from triggering Popover's onClick handler when
+          // space key is used to activate trigger.
+          // See: https://bugzilla.mozilla.org/show_bug.cgi?id=501496
+          event.key === ' ' && event.preventDefault();
+        }
+      },
+      // Custom prop getter can override all values
+      this.props.getTriggerProps
+    );
+  };
+
+  getMenuProps = (props: Object = {}) => {
+    return composePropsWithGetter(
+      {
+        // Props set by caller, e.g. DropdownContent
+        ...props,
+
+        // Props set by this component
+        id: this.getMenuId(),
+        role: 'menu'
+      },
+      // Custom prop getter can override all values
+      this.props.getMenuProps
+    );
+  };
+
+  getItemProps = (props: Object = {}, scope: Object) => {
+    const { index, item } = scope;
+    const highlightedIndex = this.getControllableValue('highlightedIndex');
+
+    return composePropsWithGetter(
+      {
+        // Props set by caller, e.g. Menu
+        ...props,
+
+        // Props set by this component
+        'aria-disabled': props.disabled,
+        id: this.getMenuItemId(index),
+        isHighlighted: highlightedIndex === index,
+        onClick: this.onItemClick.bind(null, item),
+        role: 'menuitem',
+        tabIndex: null // Unset tabIndex because we use arrow keys to navigate instead
+      },
+      // Custom prop getter can override all values
+      this.props.getItemProps,
+      scope
+    );
   };
 
   getItems = () => {
-    return this.props.data.reduce((acc, group) => {
-      return group.items && group.items.length
-        ? acc.concat(group.items.filter(item => !item.divider))
-        : acc;
-    }, []);
+    return getItems(this.props.data);
   };
 
-  onTriggerKeyDown = (event: SyntheticEvent<>) => {
-    if (event.key === 'ArrowDown') {
+  getItemIndex = (item: Item) => {
+    return this.getItems().indexOf(item);
```

Comment:
Let's not do anything now, but keep in mind that this may cause performance issues.

Reply:
Yes, it was already noted as a followup to consider storing items as an instance variable.